tests/featureWriters/kernFeatureWriter_test.py

```diff
@@ -1182,19 +1210,376 @@ def test_quantize(self, FontClass):
+
+    assert dedent(str(newFeatures)) == dedent(
+        """\
+        lookup kern_Latn {
+            lookupflag IgnoreMarks;
+            enum pos [V W] period -20;
+        } kern_Latn;
+
+        lookup kern_Nkoo {
+            lookupflag IgnoreMarks;
+            enum pos [gba-nko] period <-20 0 -20 0>;
+        } kern_Nkoo;
+
+        feature kern {
+            script latn;
+            language dflt;
+            lookup kern_Latn;
+
+            script nko;
+            language dflt;
+            lookup kern_Nkoo;
+        } kern;
+        """
+    )
+
+
+def test_kern_keep_common(FontClass):
+    """Test that if both sides are common, the output is common."""
+    glyphs = {"period": ord(".")}
+    kerning = {("period", "period"): -20}
+    ufo = makeUFO(FontClass, glyphs, None, kerning)
+    newFeatures = KernFeatureWriterTest.writeFeatures(ufo)
+
+    assert dedent(str(newFeatures)) == dedent(
+        """\
+        lookup kern_Common {
+            lookupflag IgnoreMarks;
+            pos period period -20;
+        } kern_Common;
+
+        feature kern {
+            script DFLT;
+            language dflt;
+            lookup kern_Common;
+        } kern;
+        """
+    )
+
+
+def test_kern_multi_script(FontClass):
+    """Test that glyphs with more than one script get associated with all of the
+    relevant scripts in the pair."""
+    glyphs = {"gba-nko": 0x07DC, "comma-ar": 0x060C, "lam-ar": 0x0644}
+    groups = {
+        "public.kern1.foo": ["lam-ar", "gba-nko"],
+        "public.kern2.foo": ["comma-ar"],
+    }
+    kerning = {("public.kern1.foo", "public.kern2.foo"): -20}
+    ufo = makeUFO(FontClass, glyphs, groups, kerning)
+    newFeatures = KernFeatureWriterTest.writeFeatures(ufo)
+
+    assert dedent(str(newFeatures)) == dedent(
+        """\
+        lookup kern_Arab {
+            lookupflag IgnoreMarks;
+            pos [lam-ar] [comma-ar] <-20 0 -20 0>;
+        } kern_Arab;
+
+        lookup kern_Nkoo {
+            lookupflag IgnoreMarks;
+            pos [gba-nko] [comma-ar] <-20 0 -20 0>;
+        } kern_Nkoo;
+
+        feature kern {
+            script arab;
+            language dflt;
+            lookup kern_Arab;
+
+            script nko;
+            language dflt;
+            lookup kern_Nkoo;
+        } kern;
+        """
+    )
+
+
+def test_kern_mixed_bidis(caplog, FontClass):
+    """Test that BiDi types for pairs are respected."""
+    glyphs = {
+        "a": ord("a"),
+        "comma": ord(","),
+        "alef-ar": 0x0627,
+        "comma-ar": 0x060C,
+        "one-ar": 0x0661,
+    }
+    kerning = {
+        # Undetermined: LTR
+        ("comma", "comma"): -1,
+        # LTR
+        ("a", "a"): 1,
+        ("a", "comma"): 2,
+        ("comma", "a"): 3,
+        # RTL
+        ("alef-ar", "alef-ar"): 4,
+        ("alef-ar", "comma-ar"): 5,
+        ("comma-ar", "alef-ar"): 6,
+        # Mixed: should be dropped
+        ("alef-ar", "one-ar"): 7,
+        ("one-ar", "alef-ar"): 8,
```

Comment:
so one cannot still kern arabic letters against arabic numerals... are we sure this is technically impossible to achieve? /cc @khaledhosny

Reply:
I wonder if the issue with mixed-bidi pairs such as ("alef-ar", "one-ar") or ("one-ar", "alef-ar") is the ambiguity in guessing what the font developer's intentions are as regards which is to be considered the logical "first" and "second" in the pair; or the layout engines that would itemize these into separate runs and thus the glyphs won't ever see next to one another. I remember we had several discussions about this issue but can't find a link, and wonder if anything has changed since.

Reply:
They will be in different runs. OpenType layout works on single script, single language, single direction runs, there were talks to relax this for GPOS but I’m not aware of any implementation doing this, not ones that matter anyway.

(It is possible to force them into the same run by using bidi control chars, but it would be of limited use, e.g. to kern `heh-ar.isol` with `year-ar` in a string like ؁١٤٤٤ه, though this specific case requires contextual positioning anyway.)

Reply:
thanks for the clarification